judges/pipeline.py

```diff
@@ -1,13 +1,199 @@
+
+    except requests.RequestException as error:
+        logging.info(f"Error fetching URL: {error}")
+        return pd.DataFrame([])
+
+
+# ========== FUNCTIONS: TRANSFORMING ==========
+def convert_date(datestr: str) -> datetime:
+    """Turns date string to datetime.
+    Returns date as datetime."""
+
+    datestr = datestr.split()[0]
+    length = len(datestr)
+
+    if length == 10:
+        return datetime.strptime(datestr, "%d-%m-%Y")
+    if length == 8:
+        return datetime.strptime(datestr, "%d-%m-%y")
+    if length == 9:
+        return datetime.strptime(datestr, "%d-%b-%y")
+
+
+def extract_name(judge: str, title: str) -> str:
+    """Strips away any prefix and suffix.
+    Returns a name."""
+
+    if title in judge:
+        prefix, name = judge.split(f" {title} ")
+    else:
+        for token in title.split():
+            if token in judge:
+                prefix, name = judge.split(f" {token} ")
+                break
+
+    if prefix in ["Her", "Mrs", "Ms", "Miss"]:
+        gender = "F"
+    elif prefix in ["His", "Mr"]:
+        gender = "M"
+    else:
+        gender = "X"
+
+    if "(" not in name:
+        return name.strip(), gender
+
+    name = name[:name.find("(")]
+
+    return name.strip(), gender
+
+
+def transform_df(df: pd.DataFrame,
+                 title: str,
+                 type: str,
+                 circuit: str = "") -> pd.DataFrame:
+    """Cleans and enriches the data.
+    Returns transformed data as pd.DF."""
+
+    df["appointment"] = df["appointment"].apply(convert_date)
+
+    df["judge"] = df["judge"].apply(extract_name, args=(title,))
+    df["name"] = df["judge"].str[0]
+    df["gender"] = df["judge"].str[1]
+
+    df["type"] = type
+
+    if "circuit" not in df.columns:
+        df["circuit"] = circuit
+
+    df = df[["name", "gender", "appointment", "type", "circuit"]]
+
+    return df
+
+
+def concat_dfs(dfs: list[pd.DataFrame]) -> pd.DateOffset:
+    """Concatenates pd.DFs.
+    Returns one single pd.DF"""
+
+    return pd.concat(dfs, ignore_index=True)
+
+
+# ========== FUNCTIONS: DATABASE ==========
+def get_db_connection(config) -> connect:
+    """Returns db connection."""
+
+    return connect(dbname=config["DB_NAME"],
+                   user=config["DB_USER"],
+                   password=config["DB_PASSWORD"],
+                   host=config["DB_HOST"],
+                   port=config["DB_PORT"],
+                   cursor_factory=RealDictCursor)
+
+
+def upload_data(conn: connect, df: pd.DataFrame) -> None:
+    """Insert judge data into judge table in db."""
+
+    with conn.cursor() as cur:
+        query = """
+                INSERT INTO judges
+                    (name, gender, appointed, judge_type_id, circuit_id)
+                VALUES
+                    (%s, %s, %s)
+                """
+        cur.executemany(query,
+                        [df["at"], df["site"], df["val"]])
+    conn.commit()
+
+
+# ========== MAIN ==========
 if __name__ == "__main__":
```

Comment:
Just so we can use one function rather than multiple in the "main", wrap up the script in a main().